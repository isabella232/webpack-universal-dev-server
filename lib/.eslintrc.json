{
    "extends": [
        "peerigon/flowtype",
        "peerigon/fp"
    ]
}
