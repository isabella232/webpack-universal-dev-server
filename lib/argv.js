// @flow

import path from "path";
import fs from "fs";
import yargs from "yargs";

function clarifyError(path, toWhat, err): Error {
    switch (err.code) {
        case "ENOENT":
            return new Error(`Path to ${ toWhat } does not exist: ${ path }`);
        default:
            return err;
    }
}

export default yargs
    .option("app", {
        describe: "Path to the application server executable",
        default: path.join(process.cwd(), "bin", "www"), // This is express.js' default
        coerce(appPath) {
            try {
                return fs.realpathSync(appPath);
            } catch (err) {
                throw clarifyError(appPath, "application server executable", err);
            }
        },
        alias: "a"
    })
    .option("app-args", {
        describe: "Arguments to be passed to the application server executable",
        default: ""
    })
    .option("config", {
        describe: "Path to webpack.config.js",
        default: path.join(process.cwd(), "webpack.config.js"),
        coerce(configPath) {
            try {
                return fs.realpathSync(configPath);
            } catch (err) {
                throw clarifyError(configPath, "webpack config", err);
            }
        },
        alias: "c"
    })
    .help()
    .argv;
