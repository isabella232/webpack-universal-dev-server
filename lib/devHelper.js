const fork = require("child_process").fork;
const PublishBuildStatusPlugin = require("./PublishBuildStatusPlugin");
const PauseWDSPlugin = require("./PauseWDSPlugin");

const pathToWebpackDevServer = require.resolve(".bin/webpack-dev-server");
const pathToWebpack = require.resolve(".bin/webpack");
const pathToServerProcess = require.resolve("../../bin/start-server");
const pathToWebpackConfig = require.resolve("../../config/webpack.config.js");
const APP_LISTENING = "APP/LISTENING";
const noop = Function.prototype;

function createServerProcess({
        wdsProcess,
        logServerListening = noop
    }) {
    const serverProcess = fork(
        pathToServerProcess, [],
        {
            env: process.env
        }
    );
    const messageHandlers = {
        [APP_LISTENING](message) {
            wdsProcess.send({ type: PauseWDSPlugin.CONTINUE });
            logServerListening(message.log);
        }
    };

    installMessageHandlers(serverProcess, messageHandlers);

    return serverProcess;
}

function createServerBuildProcess({
        wdsProcess,
        killServerProcess,
        createServerProcess,
        logBuildStart = noop,
        logBuildSuccess = noop,
        logBuildError = noop
    }) {
    const serverBuildProcess = fork(
        pathToWebpack, ["--config", pathToWebpackConfig, "--watch"],
        {
            silent: true,
            env: Object.assign({}, process.env, { WEBPACK_TARGET: "node" })
        }
    );
    const messageHandlers = {
        [PublishBuildStatusPlugin.COMPILATION]() {
            wdsProcess.send({ type: PauseWDSPlugin.PAUSE });
            killServerProcess();
            logBuildStart();
        },
        [PublishBuildStatusPlugin.DONE](message) {
            if (message.errorLog) {
                logBuildError(message.errorLog);
            }
            if (!message.hasErrors) {
                logBuildSuccess(message.duration);
                createServerProcess();
            }
        }
    };

    // Don't swallow errors -> awesome piping!
    serverBuildProcess.stderr.pipe(process.stderr);
    installMessageHandlers(serverBuildProcess, messageHandlers);

    return serverBuildProcess;
}

function createWDSProcess({
        logBuildStart = noop,
        logBuildSuccess = noop,
        logBuildError = noop
    }) {
    const wdsProcess = fork(
        pathToWebpackDevServer, ["--config", pathToWebpackConfig, "--hot"],
        {
            silent: true,
            env: Object.assign({}, process.env, { WEBPACK_TARGET: "web" })
        }
    );
    const messageHandlers = {
        [PublishBuildStatusPlugin.COMPILATION]() {
            logBuildStart();
        },
        [PublishBuildStatusPlugin.DONE](message) {
            if (message.errorLog) {
                logBuildError(message.errorLog);
            }
            if (!message.hasErrors) {
                logBuildSuccess(message.duration);
            }
        }
    };

    // Don't swallow errors -> awesome piping!
    wdsProcess.stderr.pipe(process.stderr);
    installMessageHandlers(wdsProcess, messageHandlers);

    return wdsProcess;
}

function installMessageHandlers(process, handlers) {
    process.on("message", message => {
        const handler = message && message.type && handlers[message.type];

        handler && handler(message);
    });
}

exports.createServerProcess = createServerProcess;
exports.createServerBuildProcess = createServerBuildProcess;
exports.createWDSProcess = createWDSProcess;
