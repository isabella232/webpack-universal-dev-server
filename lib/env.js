// @flow

const PREFIX = "WUDS_";

export const WP_CONFIG_PATH = `${ PREFIX }WP_CONFIG_PATH`;
export const PUBLISHER = `${ PREFIX }PUBLISHER`;

export function get(key: string): string {
    return String(process.env[key] || "");
}
