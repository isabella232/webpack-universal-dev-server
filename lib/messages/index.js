// @flow

import type {
    WpDevServerPauseMessage, WpDevServerResumeMessage,
    WpCompilationMessage, WpDoneMessage,
    WebpackStats
} from "../types"

export const TYPE_WP_DEV_SERVER_PAUSE = "WP_DEV_SERVER/PAUSE"
export const TYPE_WP_DEV_SERVER_RESUME = "WP_DEV_SERVER/RESUME"
export const TYPE_WP_COMPILATION = "WP/COMPILATION"
export const TYPE_WP_DONE = "WP/DONE"
export const TYPE_APP_SERVER_LISTENING = "APP_SERVER/LISTENING"

export const wpDevServerPause = (
): WpDevServerPauseMessage =>
    ({
        type: TYPE_WP_DEV_SERVER_PAUSE
    })

export const wpDevServerResume = (
): WpDevServerResumeMessage =>
    ({
        type: TYPE_WP_DEV_SERVER_RESUME
    })

export const wpCompilation = (
    publisher: string
): WpCompilationMessage =>
    ({
        type: TYPE_WP_COMPILATION,
        publisher
    })

export const wpDone = (
    publisher: string,
    stats: WebpackStats
): WpDoneMessage =>
    ({
        type: TYPE_WP_DONE,
        publisher,
        duration: stats.endTime - stats.startTime,
        hasErrors: stats.hasErrors(),
        hasWarnings: stats.hasWarnings(),
        errorLog: stats.toString("errors-only")
    })

// appServerListening message is not defined here because the message is created by the application itself
// See also our webpack-universal-tools library
