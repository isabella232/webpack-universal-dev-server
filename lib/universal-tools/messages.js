// @flow

import type {
    AppServerListeningMessage,
} from "./types";

export const TYPE_APP_SERVER_LISTENING = "APP_SERVER/LISTENING";

export const appServerListening = (
    port: number,
    host: string = "localhost"
): AppServerListeningMessage =>
    ({
        type: TYPE_APP_SERVER_LISTENING,
        port,
        host,
    });
