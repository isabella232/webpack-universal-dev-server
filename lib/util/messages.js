// @flow

const NAMESPACE = __filename;

// WDS = Webpack Dev Server (watches, compiles and serves the client bundle)
// WWP = Webpack Watch Process (watches and compiles the server bundle)
export const TYPE_WDS_PAUSE = `${ NAMESPACE }/WDS/PAUSE`;
export const TYPE_WDS_RESUME = `${ NAMESPACE }/WDS/RESUME`;
export const TYPE_WWP_COMPILATION = `${ NAMESPACE }/WWP/COMPILATION`;
export const TYPE_WWP_DONE = `${ NAMESPACE }/WWP/DONE`;

export type Message = {
    type: string;
};

export function wdsPause(): Message {
    return {
        type: TYPE_WDS_PAUSE
    };
}

export function wdsResume(): Message {
    return {
        type: TYPE_WDS_RESUME
    };
}

export function wwpCompilation(): Message {
    return {
        type: TYPE_WWP_COMPILATION
    };
}

export function wwpDone(): Message {
    return {
        type: TYPE_WWP_DONE
    };
}
