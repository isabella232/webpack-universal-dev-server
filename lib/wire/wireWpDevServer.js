// @flow

import type { Process } from "flow";
import type { ChildProcess } from "child_process";
import type { Logger, WpDoneMessage } from "../types";
import handleMessages from "../util/handleMessages";
import { buildStarted, buildSuccess, buildError } from "../util/logMessages";
import { TYPE_WP_COMPILATION, TYPE_WP_DONE } from "../util/messages";

type WireWpDevServerArgs = {
    +wpDevServer: ChildProcess,
    +process: Process,
    +log: Logger
};

export default function wireWpDevServer({
        wpDevServer,
        process,
        log
    } : WireWpDevServerArgs): void {
    const messageHandlers = new Map();

    messageHandlers.set(TYPE_WP_COMPILATION, () => {
        log(buildStarted("webpack client build"));
    });
    messageHandlers.set(TYPE_WP_DONE, (message: WpDoneMessage) => {
        if (message.errorLog) {
            log(buildError("webpack client build", message.errorLog));
        } else if (!message.hasErrors) {
            log(buildSuccess("webpack client build", message.duration));
        }
    });

    handleMessages(wpDevServer, messageHandlers);

    // Webpack errors won't reported to stderr. This is just for hard failures.
    wpDevServer.stderr.pipe(process.stderr);
}
