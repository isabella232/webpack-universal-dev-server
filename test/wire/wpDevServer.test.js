import test from "ava"
import wireWpDevServer from "../../lib/wire/wpDevServer"
import { wpCompilation, wpDone } from "../../lib/messages"
import { WritableProcess, ReadableProcess, SilentProcess } from "../helpers/fakeProcesses"
import normalizeLog from "../helpers/normalizeLog"

function noop() {}

test("should log a build started message when the wpDevServer process emitted a wpCompilation message", t => {
    const wpDevServer = new ReadableProcess()
    let receivedLog

    wireWpDevServer({
        wpDevServer,
        process,
        log(log) {
            receivedLog = log
        }
    })

    wpDevServer.emit("message", wpCompilation("test-publisher"))

    t.snapshot(normalizeLog(receivedLog))
})

test("should log a build success message when the wpDevServer process emitted a wpDone message with hasErrors false", t => {
    const wpDevServer = new ReadableProcess()
    const fakeWpStats = {
        startTime: 1000, // fake timestamps, minimum precision in milliseconds
        endTime: 2500,
        hasErrors: () => false,
        hasWarnings: noop,
        toString: noop
    }
    let receivedLog

    wireWpDevServer({
        wpDevServer,
        process,
        log(log) {
            receivedLog = log
        }
    })

    wpDevServer.emit("message", wpDone("test-publisher", fakeWpStats))

    t.snapshot(normalizeLog(receivedLog))
})

test("should log a build error message when the wpDevServer process emitted a wpDone message with an errorLog", t => {
    const wpDevServer = new ReadableProcess()
    const fakeWpStats = {
        hasErrors: () => true,
        hasWarnings: noop,
        toString: () => "Module build failed"
    }
    let receivedLog

    wireWpDevServer({
        wpDevServer,
        process,
        log(log) {
            receivedLog = log
        }
    })

    wpDevServer.emit("message", wpDone("test-publisher", fakeWpStats))

    t.snapshot(normalizeLog(receivedLog))
})

test("should pipe stderr from wpDevServer to process", async t => {
    const wpDevServer = new ReadableProcess()
    const process = new WritableProcess()

    wireWpDevServer({
        wpDevServer,
        process,
        log: noop
    })

    await new Promise((resolve, reject) => {
        wpDevServer.stderr.on("readable", resolve)
        wpDevServer.stderr.put("test", "utf8")
        wpDevServer.stderr.stop()
    })

    t.is(process.stderr.getContentsAsString("utf8"), "test")
})

test("should not throw if the wpDevServer is a silent process", t => {
    const wpDevServer = new SilentProcess()
    const process = new WritableProcess()

    wireWpDevServer({
        wpDevServer,
        process,
        log: noop
    })
})
