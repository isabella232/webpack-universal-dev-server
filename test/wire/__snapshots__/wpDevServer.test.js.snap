exports[`should log a build error message when the wpDevServer process emitted a wpDone message with an errorLog 1`] = `"[\"\",\"---------------------------------------------------------------------------------------\",\"\\u001b[31mwebpack client build error\\u001b[39m\",\"---------------------------------------------------------------------------------------\",\"\",\"Module build failed\",\"\",\"---------------------------------------------------------------------------------------\"]"`;

exports[`should log a build started message when the wpDevServer process emitted a wpCompilation message 1`] = `"[\"\\u001b[33mwebpack client build\\u001b[39m started...\"]"`;

exports[`should log a build success message when the wpDevServer process emitted a wpDone message with hasErrors false 1`] = `"[\"\\u001b[32mwebpack client build finished\\u001b[39m after 1.5s\"]"`;
